### DISCUSSION TITLE
Exclude channels in a file

### DISCUSSION THREAD

**Discussion by rhodessteve** [rusq/slackdump#217] [Q&A]
The docs suggest using the @data.txt to only download the channels you want in a text file, and to use ^ to exclude certain channels. But the combination of ^@data.txt does not work. Can we exclude a list of channels from a file? Also what is the format of the file? I have exported a CSV of channels from the Slack UI and have just the channels in that CSV.

**Comment by rusq** [comment]
Hey @rhodessteve,

Slackdump does not understand the combination of characters, as you probably already experienced, given the ^@your_file.txt syntax, treats it as the channel with the name "@your_file.txt", and you get "invalid link" error.

You probably already saw this, but for completeness:  the syntax of the file is described in [this doc](https://github.com/rusq/slackdump/blob/master/doc/usage-channels.rst#reading-data-from-the-file)

The file syntax is "one item per line", and the syntax of the file is the following:
```
# line comment 
CHANN3L
^EXCLUDED
https://slack.com/.../CHANN3L
^https://slack.com/.../CHANN3L
@nested_file.txt
```

If you have the file with excludes, you can run it through sed to add the "exclude" control character:
```
cat your_file.txt | sed 's/^/^/g' > excluded.txt
```

or, in-place:
```
sed -i.bak 's/^/^/g' your_file.txt
```

**Comment by rhodessteve** [comment]
I see. Slackdump reads the file, and the exclusions are in the file. Makes sense. Thanks for the furhter info about the formatting of the file, and that I can also nest the files. Very handy.

**Reply by rusq** [reply]
I made a note to myself to document it in details :)  Thanks for raising this.

**Comment by rusq** [comment]
I gave it a thought, and I think the approach of having ^@ introduces some unnecessary complex logic.

Consider a file is specified with negation character "^@file.txt". 
- What happens to excluded items within that file?  Should they be treated inversely?
- What if it is a nested file, should the entries from it also be excluded?
- What if it is a nested file with negation?

For example, consider the file with the following contents, let's refer to it as "first.txt":
```
CHAN1
^CHAN2
^CHAN3
@second.txt
```

`second.txt` contents:
```
CHAN4
^CHAN5
```

Using the current logic, slackdump will process:
- CHAN1
- CHAN4

and will ignore:
- CHAN2
- CHAN3
- CHAN5

If we introduce the ^@ syntax, what should happen to CHAN2,  CHAN3 and CHAN5?

**Reply by rhodessteve** [reply]
My gut feeling is that if the ^@first.txt is present on the command line then it should exclude everything in the file regardless otherwise you are introducing extra complexity and logic. Similarly if you issue a command to include @first.txt with a ^@second.txt inside it, it would also exclude all the content in the @second.txt but include all the channels in the @first, except the ones also with ^.

This gives a sort of permission hierarchy.
